### ISSUE TITLE
Create services with a load balancer

### ISSUE THREAD

**Issue by geowa4** [aws/amazon-ecs-cli#21]
ecs-cli has been wonderful. For me, it's missing the ability to create services with load balancers. As a workaround, I define tasks with ecs-cli and use aws-cli (http://docs.aws.amazon.com/cli/latest/reference/ecs/create-service.html) to create the service with the load balancer.

**Comment by veered** [comment]
+1

**Comment by MathieuLescure** [comment]
+1

**Comment by nacyot** [comment]
:+1:

**Comment by geowa4** [comment]
PSA: use the subscribe button on the right to indicate "+1".

**Comment by johndoran** [comment]
+1

**Comment by c05mic** [comment]
+1

**Comment by akatrevorjay** [comment]
Why was this closed?

**Comment by lox** [comment]
I wrote this in the meantime: https://github.com/99designs/ecs-cli

**Comment by smiller171** [comment]
Has there been any traction on this?

**Comment by andymotta** [comment]
+1

**Comment by MichaelNesterenko** [comment]
+1

**Comment by jedhassell** [comment]
+1

**Comment by khouse** [comment]
+1

**Comment by AmielYanez** [comment]
+1

**Comment by vincetse** [comment]
:+1:

**Comment by sunghoaraipark** [comment]
+1

**Comment by smiller171** [comment]
@sunghoaraipark @vincetse There is a built in voting system now using reactions. it's better to add a thumbs up reaction to the first comment

**Comment by engineeringsnap** [comment]
Hi, do we have any idea of an ETA of when / if this might be available?

**Comment by mfonsen** [comment]
As there has been no updates on this, I've been using Cloudformation for few months now for managing ECS. After recent updates, it supports all paramaters I need. Might work for some one else too.

**Comment by chrisheaththomas** [comment]
+1

**Comment by ankitcha** [comment]
+1

**Comment by kaihendry** [comment]
@geowa4 @mfonsen Would be good if you iterate the steps you use on the cli to create a cluster with a load balancer.

My current work around is to create the service manually in the console but that can fraught with error and confusion. :grin:

**Comment by ankitcha** [comment]
@kaihendry I am using following script to do what @geowa4 mentioned in the description
```
ecs-cli configure --region us-east-1 --cluster $CLUSTER_NAME

# create ecs cluster of ec2 instances
ecs-cli up --keypair $KEY_PAIR --capability-iam --size $CLUSTER_SIZE --security-group $SSH_SECURITY_GROUP --vpc $VPC_ID --subnets $SUBNET_ID --image-id $AMI_ID --instance-type $INSTANCE_TYPE --verbose

# create task definition for a docker container
ecs-cli compose --file $LANG-$ENV.yml --project-name $CLUSTER_NAME --verbose create

# using aws-cli to create elb
aws elb create-load-balancer --load-balancer-name $CLUSTER_NAME --listeners Protocol=TCP,LoadBalancerPort=8080,InstanceProtocol=TCP,InstancePort=8080 --subnets $SUBNET_ID --security-groups $ECS_SECURITY_GROUP --scheme internal

# using aws-cli (vs ecs-cli) to create service with above created task definition & elb
aws ecs create-service --service-name ecscompose-service-$CLUSTER_NAME --cluster $CLUSTER_NAME --task-definition ecscompose-simple-$CLUSTER_NAME --load-balancers $CLUSTER_NAME --desired-count 1 --deployment-configuration maximumPercent=200,minimumHealthyPercent=50
```

**Comment by leandrocr** [comment]
+1

**Comment by smiller171** [comment]
@leandrocr please use the built-in reaction system instead of a "+1" comment. GitHub finally implemented it (and you can sort issues by them) so let's use them! 😄

**Comment by TakaraShinya** [comment]
+1

**Comment by kaihendry** [comment]
Thanks for the starting point @ankitcha 

I maintain a set of scripts to setup the load balancer with ECS setup scripts here: https://github.com/kaihendry/count

I really really hope ecs-cli gets some love and fixes this _standard_ use case.

**Comment by dcworldwide** [comment]
New to ecs, dissapointed to see this use case isn't supported otb. 

Thanks for sharing your workaround @ankitcha, I'm just wondering, if you subsequently scale out/in your containers, will ELB automatically update to reflect the changes?

**Comment by ankitcha** [comment]
@dcworldwide Yes, it does work for scaling-in / out for containers.

**Comment by pierswarmers** [comment]
+1

**Comment by hscholz** [comment]
+1

**Comment by victorcoder** [comment]
FYI Support for assigning ELB/ALB is now on master, the ELB should exist before creating. https://github.com/aws/amazon-ecs-cli/pull/211

**Comment by jritsema** [comment]
Awesome @victorcoder!  Are there any docs/examples yet?

**Comment by vbudilov** [comment]
@victorcoder, I've downloaded v0.5.0 (who's releast notes specify the availability of this feature) but can't seem to figure out where to call '--target-group-arn'. The help section doesn't have it listed as one of the options for the 'up' command either. Also, the main docs weren't updated either. Are you sure that it was merged with v0.5.0?

**Comment by yinshiua** [comment]
Yes, this is merged with v0.5.0. Please look at `ecs-cli compose service up --help`

**Comment by isaacd9** [comment]
I can't seem to get this to work. I checked and double-checked that I specified `--target-group-arn`, `--container-name`, `--container-port`, and `--role` properly, but I can't get `ecs-cli` to add my service to my target group.

I ran it with `--debug` but it just seems like the load balancer info isn't being added to the service definition. Any tips on why this might be not working or getting more verbose log output from the `up` command?

(For reference I'm on `ecs-cli version 0.5.0 (1aa8028)`)

Thanks so much. :)

**Comment by isaacd9** [comment]
Ok, I think I've isolated the problem. I brought up a second cluster and LB and tried to just bring up an entirely new instance of my service on that cluster. It failed with error:

```
ERRO[0000] Error creating service                        error="InvalidParameterException: Unable to assume role and validate the specified targetGroupArn. Please verify that the ECS service role being passed has the proper permissions.\n\tstatus code: 400, request id: <request-id>" service=<servicename>
FATA[0000] InvalidParameterException: Unable to assume role and validate the specified targetGroupArn. Please verify that the ECS service role being passed has the proper permissions.```

Then, when I run `ecs-cli compose --project-name <project-name> service up ----target-group-arn <arn> --container-name nginx --container-port 80` again, the command succeeds, but fails to actually add the service to the LB. Seems like there is a bug where if the role requests 400, `ecs-cli` doesn't properly indicate an error.

As an aside, the IAM error itself is really bizarre, given that I've created and am using IAM with effectively wide open permissions. Although, perhaps, this thread isn't the proper place to discuss issues with IAM and ECS themselves.

**Comment by yinshiua** [comment]
Can you explain step by step how you run into this problem when service is not created with the LB? What commands did you run? 

> `ecs-cli compose --project-name <project-name> service up --target-group-arn <arn> --container-name nginx --container-port 80`

This should throw you an error because role is required here.

**Comment by isaacd9** [comment]
Excuse me, I just neglected to include the `role` flag, I did include it the the command. However, this does seem to be broken in that when the service has already been created, it doesn't fail without a role flag.

I'm not sure what you mean by "when the service is not created with the LB". I created the ELB and target groups from within the console and specified the ARN that I created in the command above. Do I need to do something other than just run the `ecs-cli compose` command?

**Comment by isaacd9** [comment]
Oh, huh. I was wondering why it wasn't updating when I updated the service definition, but it seems like there's no way to update the load balancer a service uses?

It looks like it's missing from `UpdateServiceInput` which `Up()` uses under the hood. Not sure how I missed this. I see it's in the docs now. So sorry. :(

**Comment by yinshiua** [comment]
Ahh, I see. The behavior is expected, see [documentation](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-load-balancing.html):
> After you create a service, the target group ARN or load balancer name, container name, and container port specified in the service definition are immutable. You cannot add, remove, or change the load balancer configuration of an existing service. If you update the task definition for the service, the container name and container port that were specified when the service was created must remain in the task definition.

To avoid confusions here, we should add validations on cli when the service already exists.

**Comment by isaacd9** [comment]
Yeah, this was really confusing and spent about an hour on it last night. An error message would be nice. :)

**Comment by yinshiua** [comment]
Sorry for the confusion, we definitely need better error handling here. I will create a new issue to follow up on this :) Thanks!

**Comment by shatil** [comment]
@yinshiua, any idea whether this will be created using Cloudformation or via API?

**Comment by yinshiua** [comment]
@shatil Currently, we only support using existing ELB in Create Service. We are planning on supporting the ELB creation possibly using the CloudFormation, but I do not have an ETA yet.

**Comment by kaihendry** [comment]
Since I struggle with this every week, I thought I would share the script I use to deploy services with.
https://github.com/unee-t/bugzilla/blob/master/deploy.sh

Prerequisites:
1. Create the ALB & targets manually in the console
2. I assume you've setup *ecsServiceRole* with the **AmazonEC2ContainerServiceRole** policy

So I wish there was something to quickly create the ALBs, not easy thanks to a little more than often certificate dance.

And not sure what or where the role was supposed to be created.

**Comment by mhaggerty042** [comment]
@kaihendry I get a 404 on your link

**Comment by kaihendry** [comment]
Oops, @mhaggerty042 
https://github.com/unee-t/bugzilla-customisation/blob/master/deploy.sh

**Comment by jasonrhaas** [comment]
I see that the latest docs have already been updated with support for adding a load balancer to a service -- https://docs.aws.amazon.com/AmazonECS/latest/developerguide/cmd-ecs-cli-compose-service.html

Any idea when this version will be released?  Are there instructions with how to build from source?

**Comment by jordantomax** [comment]
@SoManyHs Is what @jasonrhaas says true? Or is the best way still to generate tasks using `ecs-cli compose` and use `aws ecs create-service` with the new task definition?

**Comment by SoManyHs** [comment]
@jordantomax If you wanted to create a load balancer in the same workflow as creating an ECS service (rather than specifying one that was created separately), you might be interested in checking out the [CDK](https://github.com/awslabs/aws-cdk)! https://awslabs.github.io/aws-cdk/refs/_aws-cdk_aws-ecs.html#include-a-load-balancer

**Comment by jordantomax** [comment]
@PettitWesley Sorry to re-open. I'm happy to create a new issue if you think that's the best thing to do. However, the command you listed still isn't working for me :(. The docs you sent are for `create` not for `up`. Was that your intention? Thanks in advance.

**Comment by PettitWesley** [comment]
@jordantomax Yeah, please create a new issue if you want the ECS CLI to support creating a load balancer, as opposed to merely creating an ECS Service with an existing load balancer. 

Sorry about the docs link- I've edited it to include the link for `up`. However, `up` works the same as `create`, it just additionally sets the desired count of the service to 1. 

Also, for your problem with the command- can you create a new issue and show the exact command that you ran, etc? That would help us determine your problem. Right now, our best guess is that you might not be using the correct ARN with the target group flag- that's something we've seen before.

**Comment by jordantomax** [comment]
@PettitWesley Got it. Actually looks like my problem was that I was specifying the flag on the `compose` command instead of on the `up` command. All working now, thank you!

**Comment by pecigonzalo** [comment]
Will this parameter be added to the ecs-params.yml?

**Comment by PettitWesley** [comment]
@pecigonzalo The parameter can definitely be added to the ECS Params yaml file! Can you create an issue so that we can track that request? Also, we always welcome pull requests, especially for small improvements like this :)

**Comment by pecigonzalo** [comment]
Sure things, I was unsure if it was support and just not documented. Will
open an issue and try to PR the fix.

**Comment by PettitWesley** [comment]
@pecigonzalo Awesome!
